File: FairPlay_Streaming_Server_SDK_5.1 2/Development/Key_Server_Module/Swift/Sources/src/base/structures/base_fps_structures.swift
//
// base_fps_structures.swift: Contains the structure definitions for the Base class.
//
// Copyright © 2023-2025 Apple Inc. All rights reserved.
//

import Foundation

/// Base container where common code is implemented.
public struct Base: Codable {
  public var fpsOperations: FPSOperations

  enum CodingKeys: String, CodingKey {
    case fpsOperations = "fairplay-streaming-request"
  }

  public init() {
    fpsOperations = FPSOperations.init()
  }
}

/// Information about all create-ckc operations in a request.
///
/// Contains a vector of FPSOperation. This is necessary if receiving multiple requests in the same JSON.
public struct FPSOperations: Codable {
  var version: UInt32
  var operationsPtr: [FPSOperation]

  enum CodingKeys: String, CodingKey {
    case version
    case operationsPtr = "create-ckc"
  }

  public init() {
    version = 0
    operationsPtr = []
  }
}

/// Information about a single create-ckc operation.
///
/// This is the basic structure of a FairPlay Streaming key request after the JSON has been parsed.
public struct FPSOperation: Codable {

  var id: UInt64
  var spc: [UInt8]
  /// True when input SPC is a SyncSPC with check-in
  var isCheckIn: Bool
  var assetInfo: AssetInfo

  // Extension
  public var ext: FPSOperationExtension  // room for extension values if not in use template to ()

  enum CodingKeys: String, CodingKey {
    case id
    case spc
    case isCheckIn = "check-in"
    case assetInfo = "asset-info"
  }

  public init(from decoder: Decoder) throws {
    // Save any parsing errors until the end instead of throwing them immediately.
    // This way we can print all errors if there are more than one.
    var status = FPSStatus.noErr

    let values = try decoder.container(keyedBy: CodingKeys.self)

    // ID - optional, defaults to 0
    do {
      id = try values.decodeIfPresent(UInt64.self, forKey: .id) ?? 0
    } catch _ as DecodingError {
      do {
        // Try again as a string instead
        id = UInt64(try values.decodeIfPresent(String.self, forKey: .id)!) ?? 0
      } catch _ as DecodingError {
        id = 0 // default value if not provided
      }
    }

    // SPC - required
    spc = [0]
    let base64Request = try values.decode(String.self, forKey: .spc)
    if let decodedStr = Data(base64Encoded: base64Request) {
      spc = [UInt8](decodedStr)
    } else {
      fpsLogError(FPSStatus.parserErr, "Error parsing HDCP type:err")
      status = FPSStatus.parserErr
    }

    isCheckIn = try values.decodeIfPresent(Bool.self, forKey: .isCheckIn) ?? false

    // Extension
    ext = FPSOperationExtension()

    // Parse asset-info
    let assetInfoArray: [AssetInfo] = try values.decodeIfPresent(Array.self, forKey: .assetInfo) ?? []

    if !assetInfoArray.isEmpty {
      // We only handle one asset per request. Use the first one in the json array.
      assetInfo = assetInfoArray[0]
    } else {
      // Asset info not provided in input. They should be provided later.
      assetInfo = AssetInfo()
    }

    // Custom operations
    do {
      try parseCreateCKCOperationCustom(&self, decoder)
    } catch let error as FPSStatus {
      fpsLogError(error, "parseCreateCKCOperationCustom failed")
      status = error
    }

    if status != FPSStatus.noErr {
      throw status
    }
  }

  public func encode(to encoder: Encoder) throws {
  }
}

/// Protection requirements related to a particular asset.
public struct AssetInfo: Codable {

  var key: [UInt8]
  var iv: [UInt8]
  var isCKProvided: Bool

  public var hdcpReq: base_constants.FPSHDCPRequirement  // one of the FPSHDCPRequirement enums

  // Expirations
  var leaseDuration: UInt32     // Lease duration (starts at SPC creation time)
  var rentalDuration: UInt32    // rental duration in seconds. Starts at asset download time
  var playbackDuration: UInt32  // playback duration in seconds. Starts at asset first playback time

  // Offline HLS parameters
  var licenseType: base_constants.FPSLicenseType
  var streamId: [UInt8]   // unique Id of each HLS sub-stream
  var titleId: [UInt8]    // Id of a title (program). Same for all HLS subs-tream of a give title.

  // Extension
  public var ext: AssetInfoExtension

  enum CodingKeys: String, CodingKey {
    case key = "content-key"
    case iv = "content-iv"
    case isCKProvided
    case leaseDuration = "lease-duration"
    case hdcpReq = "hdcp-type"
    case offlinehls = "offline-hls"
  }

  enum OfflineHLSKeys: String, CodingKey {
    case licenseType
    case rentalDuration = "rental-duration"
    case playbackDuration = "playback-duration"
    case streamId = "stream-id"
    case titleId = "title-id"
  }

  public init() {
    key = [UInt8](repeating: 0, count: base_constants.AES128_KEY_SZ)
    iv = [UInt8](repeating: 0, count: base_constants.AES128_IV_SZ)
    isCKProvided = false
    leaseDuration = 0
    rentalDuration = 0
    playbackDuration = 0

    hdcpReq = base_constants.FPSHDCPRequirement.hdcpType0

    licenseType = base_constants.FPSLicenseType.none
    streamId = []
    titleId = []

    ext = AssetInfoExtension()
  }

  public init(from decoder: Decoder) throws {
    // Save any parsing errors until the end instead of throwing them immediately.
    // This way we can print all errors if there are more than one.
    var status = FPSStatus.noErr

    let values = try decoder.container(keyedBy: CodingKeys.self)

    isCKProvided = true

    // content key - optional for lease renewals
    key = Array(repeating: 0, count: base_constants.AES128_KEY_SZ)
    if var contentKey = try values.decodeIfPresent(String.self, forKey: .key) {
      if !contentKey.isEmpty {
        // Remove any initial 0x at the front
        if contentKey.hasPrefix("0x") {
          contentKey.removeFirst(2)
        }

        // Array fromHexString requires an even number of characters, so insert a 0 at the front if odd.
        if contentKey.count % 2 == 1 {
          contentKey.insert("0", at: contentKey.startIndex)
        }

        key = [UInt8](fromHexString: contentKey)!
      }
      key += [UInt8](repeating: 0, count: (base_constants.AES128_KEY_SZ - key.count))
    } else {
      isCKProvided = false
    }

    // content iv - optional for lease renewals
    iv = Array(repeating: 0, count: base_constants.AES128_IV_SZ)
    if var contentIV = try values.decodeIfPresent(String.self, forKey: .iv) {
      if !contentIV.isEmpty {
        // Remove any initial 0x at the front
        if contentIV.hasPrefix("0x") {
          contentIV.removeFirst(2)
        }

        // Array fromHexString requires an even number of characters, so insert a 0 at the front if odd.
        if contentIV.count % 2 == 1 {
          contentIV.insert("0", at: contentIV.startIndex)
        }

        iv = [UInt8](fromHexString: contentIV)!
      }
      iv += [UInt8](repeating: 0, count: (base_constants.AES128_IV_SZ - iv.count))
    } else {
      isCKProvided = false
    }

    leaseDuration =
      try values.decodeIfPresent(UInt32.self, forKey: .leaseDuration)
      ?? base_constants.NO_LEASE_DURATION

    // Offline HLS parameters
    licenseType = base_constants.FPSLicenseType.none
    streamId = Array()
    titleId = Array()
    rentalDuration = 0
    playbackDuration = 0

    // Extension
    ext = AssetInfoExtension()

    // HDCP requirement (-1: not required, 0: type 0, 1: type 1)
    hdcpReq = base_constants.FPSHDCPRequirement.hdcpNotRequired
    if values.contains(.hdcpReq) {
      let hdcpType = try values.decode(Int32.self, forKey: .hdcpReq)

      do {
        hdcpReq = try AssetInfo.parseHDCPType(hdcpType)
      } catch let error as FPSStatus {
        fpsLogError(error, "Error parsing HDCP type: \(hdcpType)")
        status = error
      }
    } else {
      Log.debug("Warning! HDCP Type not provided, defaulting to Type 0")
      hdcpReq = base_constants.FPSHDCPRequirement.hdcpType0
    }

    // Offline HLS parameters
    if values.contains(.offlinehls) {
      licenseType = base_constants.FPSLicenseType.offlineHLS
      do {
        try parseOfflineHLS(decoder)
      } catch let error as FPSStatus {
        fpsLogError(error, "parseOfflineHLS failed")
        status = error
      }
    }

    // Custom operations
    do {
      try parseAssetInfoCustom(&self, decoder)
    } catch let error as FPSStatus {
      fpsLogError(error, "parseAssetInfoCustom failed")
      status = error
    }

    if status != FPSStatus.noErr {
      throw status
    }
  }

  public func encode(to encoder: Encoder) throws {
  }
}

/// Return data for a single create-ckc operation (expected to be returned in the output JSON).
public struct FPSResult: Codable {
  public var id: UInt64
  public var status: FPSStatus
  public var hu: [UInt8]
  public var ckc: [UInt8]

  public var sessionId: UInt64  // Parsed from Reference Time Tag TLLV

  // Sync TLLV
  public var isCheckIn: Bool
  public var syncServerChallenge: UInt64
  public var syncFlags: UInt64
  public var syncTitleId: [UInt8]
  public var durationToRentalExpiry: UInt32
  public var recordsDeleted: Int  // number of keys deleted as reported by check-in variant of SyncTLLV
  public var deletedContentIDs: [UInt8]

  // Device Identity Data
  public var deviceIdentitySet: Bool
  public var fpdiVersion: UInt32
  public var deviceClass: UInt32
  public var vendorHash: [UInt8]
  public var productHash: [UInt8]
  public var fpVersionREE: UInt32
  public var fpVersionTEE: UInt32
  public var osVersion: UInt32

  public var vmDeviceInfo: Optional<VMDeviceInfo>

  // Extension
  public var ext: FPSResultExtension  // room for extension values, if not in use template to ()

  // Coding keys, raw values for enum cases must be literals
  enum CodingKeys: String, CodingKey {
    case id
    case status
    case hu
    case ckc
    case icCheckIn = "check-in-server-challenge"
    case syncServerChallenge
    case syncFlags = "check-in-flags"
    case syncTitleId = "check-in-title-id"
    case durationToRentalExpiry = "duration-left"
    case recordsDeleted
    case deletedContentIDs = "check-in-stream-id"
    case deviceIdentitySet
    case fpdiVersion = "fpdi-version"
    case deviceClass = "device-class"
    case vendorHash = "vendor-hash"
    case productHash = "product-hash"
    case fpVersionREE = "fps-ree-version"
    case fpVersionTEE = "fps-tee-version"
    case osVersion = "os-version"
    case hostDeviceClass = "host-device-class"
    case hostOSVersion = "host-os-version"
    case hostVMProtocolVersion = "host-vm-protocol-version"
    case guestDeviceClass = "guest-device-class"
    case guestOSVersion = "guest-os-version"
    case guestVMProtocolVersion = "guest-vm-protocol-version"
  }

  public init() {
    id = 0
    status = FPSStatus.noErr
    hu = Array(repeating: 0, count: base_constants.FPS_V1_HU_SZ)
    ckc = [UInt8]()

    sessionId = 0

    isCheckIn = false
    syncServerChallenge = 0
    syncFlags = 0
    syncTitleId = Array(repeating: 0, count: base_constants.FPS_MAX_TITLE_ID_LENGTH)
    durationToRentalExpiry = 0
    recordsDeleted = 0  // number of keys deleted as reported by check-in variant of SyncTLLV
    deletedContentIDs = [UInt8]()

    deviceIdentitySet = false
    fpdiVersion = 0
    deviceClass = 0
    vendorHash = Array(repeating: 0, count: base_constants.FPS_VENDOR_HASH_SIZE)
    productHash = Array(repeating: 0, count: base_constants.FPS_PRODUCT_HASH_SIZE)
    fpVersionREE = 0
    fpVersionTEE = 0
    osVersion = 0
    vmDeviceInfo = Optional.none


    ext = FPSResultExtension()
  }

  public init(from decoder: any Decoder) throws {
    self.init()
  }

  public func encode(to encoder: Encoder) throws {
    var json = encoder.container(keyedBy: CodingKeys.self)

    var huHex: String = ""
    for byte in hu {
      huHex += String(format: "%02X", byte)
    }
    try json.encode(id, forKey: .id)
    try json.encode(status.rawValue, forKey: .status)

    if status == FPSStatus.noErr {
      try json.encode(huHex, forKey: .hu)

      if isCheckIn {
        let sync_str = String(syncServerChallenge)
        try json.encode(sync_str, forKey: .syncServerChallenge)
      }

      if syncFlags != 0 {
        try json.encode(String(format: "%X", syncFlags), forKey: .syncFlags)
        try json.encode(String(durationToRentalExpiry), forKey: .durationToRentalExpiry)

        var syncTitleIdHex = ""
        for byte in syncTitleId {
          syncTitleIdHex += String(format: "%02X", byte)
        }
        try json.encode(syncTitleIdHex, forKey: .syncTitleId)

        if recordsDeleted > 0 && deletedContentIDs.count != 0 {
          var checkInContentIds: [String] = []
          for i in 0...recordsDeleted {
            let index = i * base_constants.FPS_OFFLINE_CONTENTID_LENGTH
            var hexContentId = ""
            for byte in deletedContentIDs[
              index...(index + base_constants.FPS_OFFLINE_CONTENTID_LENGTH)]
            {
              hexContentId += String(format: "%02X", byte)
            }
            checkInContentIds.append(hexContentId)
          }
          try json.encode(checkInContentIds, forKey: .deletedContentIDs)
        }
      }

      if deviceIdentitySet {
        try json.encode(fpdiVersion, forKey: .fpdiVersion)
        try json.encode(deviceClass, forKey: .deviceClass)
        try json.encode(vendorHash.map { String(format: "%02X", $0) }.joined(), forKey: .vendorHash)
        try json.encode(productHash.map { String(format: "%02X", $0) }.joined(), forKey: .productHash)
        try json.encode(String(format: "%08X", fpVersionREE), forKey: .fpVersionREE)
        try json.encode(String(format: "%08X", fpVersionTEE), forKey: .fpVersionTEE)
        try json.encode(String(format: "%08X", osVersion), forKey: .osVersion)
      }

      if let vmDeviceInfo {
          // Print Host VM Information
          try json.encode(vmDeviceInfo.hostDeviceClass.description, forKey: .hostDeviceClass)
          try json.encode(String(format: "%08X", vmDeviceInfo.hostOSVersion), forKey: .hostOSVersion)
          try json.encode(vmDeviceInfo.hostDeviceClass.description, forKey: .hostDeviceClass)
          try json.encode(vmDeviceInfo.hostVMProtocolVersion, forKey: .hostVMProtocolVersion)

          // Print Guest VM Information
          try json.encode(vmDeviceInfo.guestDeviceClass.description, forKey: .guestDeviceClass)
          try json.encode(String(format: "%08X", vmDeviceInfo.guestOSVersion), forKey: .guestOSVersion)
          try json.encode(vmDeviceInfo.guestDeviceClass.description, forKey: .guestDeviceClass)
          try json.encode(vmDeviceInfo.guestVMProtocolVersion, forKey: .guestVMProtocolVersion)
      }

      try serializeCKCNodeCustom(encoder: encoder)

      if !isCheckIn && ckc.count > 0 {
        let ckcData = Data(bytes: ckc, count: ckc.count)
        let base64Ckc = ckcData.base64EncodedString()
        try json.encode(base64Ckc, forKey: .ckc)
      }
    }
  }
}

/// Return data for all create-ckc operations in a request.
///
/// Contains a vector of FPSResult. This is necessary if multiple requests are sent at once (much like FPSOperations).
public struct FPSResults: Codable {
  public var resultPtr: [FPSResult]

  // Extension
  public var ext: FPSResultsExtension

  enum CodingKeys: String, CodingKey {
    case fpsResults = "fairplay-streaming-response"
    case resultPtr = "create-ckc"
  }

  public init() {
    resultPtr = [FPSResult]()
    ext = FPSResultsExtension()
  }

  public init(from decoder: any Decoder) throws {
    self.init()
  }

  public func encode(to encoder: any Encoder) throws {
    var json = encoder.container(keyedBy: CodingKeys.self)
    var response_container = json.nestedContainer(keyedBy: CodingKeys.self, forKey: .fpsResults)

    try response_container.encode(resultPtr, forKey: .resultPtr)

    try serializeResultsCustom(encoder: encoder)
  }
}

File: FairPlay_Streaming_Server_SDK_5.1 2/Development/Key_Server_Module/Rust/src/base/structures/base_fps_structures.rs
//
// base_fps_structures.rs: Contains the structure definitions for the Base class.
//
// Copyright © 2023-2025 Apple Inc. All rights reserved.
//

use crate::base::base_constants;
use crate::base::base_constants::FPSHDCPRequirement;
use crate::base::base_constants::FPS_V1_HU_SZ;
use super::base_server_structures::VMDeviceInfo;
use crate::extension_structures;
use crate::validate::FPSStatus;
use std::fmt::Debug;

/// Base container where common code is implemented.
pub struct Base {}

/// Information about all create-ckc operations in a request.
///
/// Contains a vector of FPSOperation. This is necessary if receiving multiple requests in the same JSON.
#[derive(Debug, Default)]
pub struct FPSOperations {
    pub operationsPtr: Vec<FPSOperation>,
}

/// Information about a single create-ckc operation.
///
/// This is the basic structure of a FairPlay Streaming key request after the JSON has been parsed.
#[derive(Debug, Default, Clone)]
pub struct FPSOperation {
    pub id: u64,
    pub spc: Vec<u8>,
    /// True when input SPC is a SyncSPC with check-in
    pub isCheckIn: bool,
    pub assetInfo: AssetInfo,

    // Extension
    pub extension: extension_structures::FPSOperationExtension,
}

/// Protection requirements related to a particular asset.
#[derive(Debug, Clone)]
pub struct AssetInfo {
    pub key: Vec<u8>,
    pub iv: Vec<u8>,
    pub isCKProvided: bool,     // true if key and iv are valid

    pub hdcpReq: u64,           // one of the FPSHDCPRequirement enums. Using type u64 so we can test with invalid values.

    // Expirations
    pub leaseDuration: u32,     // Lease duration (starts at SPC creation time)
    pub rentalDuration: u32,    // rental duration in seconds. Starts at asset download time
    pub playbackDuration: u32,  // playback duration in seconds. Starts at asset first playback time

    // Offline HLS parameters
    pub licenseType: u32,
    pub streamId: Option<Vec<u8>>,      // unique Id of each HLS sub-stream
    pub titleId: Option<Vec<u8>>,       // Id of a title (program). Same for all HLS substreams of a give title.

    // Extension
    pub extension: extension_structures::AssetInfoExtension,
}

impl Default for AssetInfo {
    fn default() -> AssetInfo {
        AssetInfo {
            isCKProvided: false,
            key: vec![0; base_constants::AES128_KEY_SZ],
            iv: vec![0; base_constants::AES128_IV_SZ],
            leaseDuration: 0,
            rentalDuration: 0,
            playbackDuration: 0,
            hdcpReq: FPSHDCPRequirement::hdcpNotRequired as u64,

            licenseType: 0,
            streamId: None,
            titleId: None,

            extension: Default::default(),
        }
    }
}

/// Return data for a single create-ckc operation (expected to be returned in the output JSON).
#[derive(Debug)]
pub struct FPSResult {
    pub id: u64,
    pub status: FPSStatus,
    pub hu: Vec<u8>,
    pub ckc: Vec<u8>,

    pub sessionId: u64, // Parsed from Reference Time Tag TLLV

    // Sync TLLV
    pub isCheckIn: bool,
    pub syncServerChallenge: u64,
    pub syncFlags: u64,
    pub syncTitleId: Vec<u8>,
    pub durationToRentalExpiry: u32,
    pub recordsDeleted: usize, // number of keys deleted as reported by check-in variant of SyncTLLV
    pub deletedContentIDs: Vec<u8>,

    // Device Identity Data
    pub deviceIdentitySet: bool,
    pub fpdiVersion: u32,
    pub deviceClass: u32,
    pub vendorHash: Vec<u8>,
    pub productHash: Vec<u8>,
    pub fpVersionREE: u32,
    pub fpVersionTEE: u32,
    pub osVersion: u32,
    pub vmDeviceInfo: Option<VMDeviceInfo>,

    // Extension
    pub extension: extension_structures::FPSResultExtension, // room for extension values, if not in use template to ()
}

impl Default for FPSResult {
    fn default() -> FPSResult {
        FPSResult {
            id: 0,
            status: FPSStatus::noErr,
            hu: vec![0; FPS_V1_HU_SZ],
            ckc: Vec::new(),

            sessionId: 0,

            isCheckIn: false,
            syncServerChallenge: 0,
            syncFlags: 0,
            syncTitleId: vec![0; base_constants::FPS_MAX_TITLE_ID_LENGTH],
            durationToRentalExpiry: 0,
            recordsDeleted: 0, // number of keys deleted as reported by check-in variant of SyncTLLV
            deletedContentIDs: Vec::new(),

            deviceIdentitySet: false,
            fpdiVersion: 0,
            deviceClass: 0,
            vendorHash: vec![0; base_constants::FPS_VENDOR_HASH_SIZE],
            productHash: vec![0; base_constants::FPS_PRODUCT_HASH_SIZE],
            fpVersionREE: 0,
            fpVersionTEE: 0,
            osVersion: 0,
            vmDeviceInfo: None,

            extension: Default::default(),
        }
    }
}

/// Return data for all create-ckc operations in a request.
///
/// Contains a vector of FPSResult. This is necessary if multiple requests are sent at once (much like FPSOperations).
#[derive(Debug, Default)]
pub struct FPSResults {
    pub resultPtr: Vec<FPSResult>,

    // Extension
    pub extension: extension_structures::FPSResultsExtension, // room for extension values, if not in use template to ()
}

/// Structure used to call C library function `PartnerKSMCreateKeyPayload`.
///
/// All members are 64B size (u64 or pointer) for easier compatibility.
#[derive(Debug)]
#[repr(C)]
pub struct KSMKeyPayload {
    pub version: u64,                       // in: version of the structure. Currently supported version is 1
    pub contentKey: *const u8,              // in: content key
    pub contentKeyLength: u64,              // in: only 16 byte long keys accepted at the moment
    pub contentIV: *const u8,               // in: content IV
    pub contentIVLength: u64,               // in: must be 16 bytes
    pub contentType: u64,                   // in: one of KSMKeyPayloadContentType enums
    pub SK_R1: *const u8,                   // in: content of FPSTLLVTagValue::sessionKeyR1Tag (0x3d1a10b8bffac2ec) TLLV
    pub SK_R1Length: u64,                   // in: size of SK_R1 data
    pub R2: *const u8,                      // in: content of FPSTLLVTagValue::r2tag (0x71b5595ac1521133) TLLV
    pub R2Length: u64,                      // in: size of R2 data
    pub R1Integrity: *const u8,             // in: content of FPSTLLVTagValue::sessionKeyR1IntegrityTag (0xb349d4809e910687) TLLV
    pub R1IntegrityLength: u64,             // in: size of R1 integrity data
    pub supportedKeyFormats: *const u64,    // in: either delivered in FPSTLLVTagValue::supportedKeyFormatTag (0x8d8e84fa6cc35eb7) or set to 16 byte key (FPSKeyFormatTag::buf16Byte) for older devices
    pub numberOfSupportedKeyFormats: u64,   // in: number of key formats in supportedKeyFormats array
    pub cryptoVersionUsed: u64,             // in: delivered in FPSTLLVTagValue::protocolVersionUsedTag (0x5d81bcbcc7f61703) TLLV
    pub provisioningData: *const u8,        // in: server Provisioning Data generated by WWDR
    pub provisioningDataLength: u64,        // in: size of the Provisioning Data
    pub certHash: *const u8,                // in: server certificate hash delivered in SPC header
    pub certHashLength: u64,                // in: size of the certificate hash
    pub clientHU: *const u8,                // out: client HU. Memory allocated by caller. Allocated buffer size should be passed in clientHULength
    pub clientHULength: u64,                // in/out: in: buffer size allocated for clientHU. out: actual HU size returned
    pub contentKeyTLLVTag: u64,             // out: Content Key TLLV tag to send to the client device
    pub contentKeyTLLVPayload: *mut u8,     // out: payload of Content Key TLLV. Memory allocated by caller. Allocated buffer size should be passed in contentKeyTLLVPayloadLength
    pub contentKeyTLLVPayloadLength: u64,   // in/out: in: buffer size allocated for contentKeyTLLVPayload. out: actual content key TLLV payload size returned
    pub R1: *mut u8,                        // out: R1 data. Memory allocated by caller. Allocated buffer size should be passed in R1Length
    pub R1Length: u64,                      // in/out: in: buffer size allocated for R1. out: actual R1 size returned
}
